File: devtools/fixwin.c
#define _GNU_SOURCE
// fixwin — a fixed-size xdg-toplevel (min == max) for placement tests: the
// dialog/splash shape (a Discord-updater-window stand-in).
// usage: fixwin <w> <h> [title]
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <unistd.h>

#include "fixwin-protocol.h"
#include <wayland-client.h>

static struct wl_display   *g_dpy;
static struct wl_compositor *g_comp;
static struct wl_shm       *g_shm;
static struct xdg_wm_base  *g_wm;
static struct wl_surface   *g_surf;
static struct xdg_toplevel *g_top;
static struct wl_buffer    *g_buf;
static int                  g_w = 310, g_h = 360, g_done;
static const char          *g_title = "fixwin";

static void reg_bind(void *d, struct wl_registry *r, uint32_t id, const char *ifc, uint32_t ver) {
    (void) d;
    (void) ver;
    if (!strcmp(ifc, "wl_compositor") && !g_comp)
        g_comp = wl_registry_bind(r, id, &wl_compositor_interface, 4);
    else if (!strcmp(ifc, "wl_shm") && !g_shm)
        g_shm = wl_registry_bind(r, id, &wl_shm_interface, 1);
    else if (!strcmp(ifc, "xdg_wm_base") && !g_wm)
        g_wm = wl_registry_bind(r, id, &xdg_wm_base_interface, 1);
}
static void reg_remove(void *d, struct wl_registry *r, uint32_t id) {
    (void) d;
    (void) r;
    (void) id;
}
static const struct wl_registry_listener regl = { reg_bind, reg_remove };

static void wm_ping(void *d, struct xdg_wm_base *b, uint32_t s) {
    (void) d;
    xdg_wm_base_pong(b, s);
}
static const struct xdg_wm_base_listener wml = { .ping = wm_ping };

static void top_configure(void *d, struct xdg_toplevel *t, int32_t w, int32_t h, struct wl_array *s) {
    (void) d;
    (void) t;
    (void) w;
    (void) h;
    (void) s;
}
static void top_close(void *d, struct xdg_toplevel *t) {
    (void) d;
    (void) t;
    g_done = 1;
}
static const struct xdg_toplevel_listener topl = { .configure = top_configure, .close = top_close };

int main(int argc, char **argv) {
    g_w = argc > 1 ? atoi(argv[1]) : 310;
    g_h = argc > 2 ? atoi(argv[2]) : 360;
    if (g_w < 1)
        g_w = 1;
    if (g_h < 1)
        g_h = 1;
    g_title = argc > 3 ? argv[3] : "fixwin";

    g_dpy = wl_display_connect(NULL);
    if (!g_dpy)
        return 1;

    struct wl_registry *reg = wl_display_get_registry(g_dpy);
    wl_registry_add_listener(reg, &regl, NULL);
    wl_display_dispatch(g_dpy);
    wl_display_roundtrip(g_dpy);

    g_surf = wl_compositor_create_surface(g_comp);
    struct xdg_surface *xsd = xdg_wm_base_get_xdg_surface(g_wm, g_surf);
    g_top                = xdg_surface_get_toplevel(xsd);
    xdg_wm_base_add_listener(g_wm, &wml, NULL);
    xdg_toplevel_add_listener(g_top, &topl, NULL);
    xdg_toplevel_set_app_id(g_top, g_title);
    xdg_toplevel_set_title(g_top, g_title);
    xdg_toplevel_set_min_size(g_top, g_w, g_h);
    xdg_toplevel_set_max_size(g_top, g_w, g_h); // min == max: the dialog shape

    // the fd lives with the process; one fixture window, no reuse
    const size_t N  = (size_t)g_w * g_h * 4;
    int          fd = memfd_create("fixwin", 0);
    ftruncate(fd, N);
    uint32_t   *px = mmap(NULL, N, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    for (size_t i = 0; i < (size_t)g_w * g_h; i++)
        px[i] = 0xff202030;
    struct wl_shm_pool *pool = wl_shm_create_pool(g_shm, fd, N);
    g_buf                    = wl_shm_pool_create_buffer(pool, 0, g_w, g_h, g_w * 4, WL_SHM_FORMAT_ARGB8888);
    wl_shm_pool_destroy(pool);

    wl_surface_attach(g_surf, g_buf, 0, 0);
    wl_surface_commit(g_surf);

    while (!g_done)
        if (wl_display_dispatch(g_dpy) < 0)
            break;

    xdg_toplevel_destroy(g_top);
    wl_surface_destroy(g_surf);
    wl_buffer_destroy(g_buf);
    xdg_wm_base_destroy(g_wm);
    wl_shm_destroy(g_shm);
    wl_compositor_destroy(g_comp);
    wl_registry_destroy(reg);
    wl_display_disconnect(g_dpy);
    munmap(px, N);
    close(fd);
    return 0;
}

File: devtools/src/bin/fixwin.rs
//! fixwin — a fixed-size xdg-toplevel (min == max) for placement tests: the
//! dialog/splash shape (a Discord-updater-window stand-in).
//!
//! usage: fixwin <w> <h> [title]
//!
//! Rust port of the C fixture (awesome-rust plan, Track D).

use std::os::fd::AsFd;

use rustix::fs::{MemfdFlags, memfd_create};
use rustix::io::write;
use wayland_client::protocol::{
    wl_buffer, wl_compositor, wl_registry, wl_shm, wl_shm_pool, wl_surface,
};
use wayland_client::{Connection, Dispatch, QueueHandle, delegate_noop};
use wayland_protocols::xdg::shell::client::{
    xdg_surface::{Event as XdgSurfaceEvent, XdgSurface},
    xdg_toplevel::{Event as XdgToplevelEvent, XdgToplevel},
    xdg_wm_base::{Event as XdgWmBaseEvent, XdgWmBase},
};

// queue state: one fixture window, no reuse. The globals are collected first
// (C drains the whole registry before creating anything); the window is built
// and committed exactly once they are all in, so the xdg toplevel exists
// BEFORE the first buffer commit.
struct State {
    width: i32,
    height: i32,
    title: String,
    comp: Option<wl_compositor::WlCompositor>,
    shm: Option<wl_shm::WlShm>,
    wm: Option<XdgWmBase>,
    surf: Option<wl_surface::WlSurface>,
    buf: Option<wl_buffer::WlBuffer>,
    done: bool,
}

impl State {
    fn new(width: i32, height: i32, title: String) -> Self {
        Self {
            width,
            height,
            title,
            comp: None,
            shm: None,
            wm: None,
            surf: None,
            buf: None,
            done: false,
        }
    }

    fn ready(&self) -> bool {
        self.comp.is_some() && self.shm.is_some() && self.wm.is_some()
    }

    // C order: surface -> xdg surface -> toplevel (+ app id, title, min ==
    // max) -> pool/buffer -> attach -> commit. Never ack configure: the first
    // committed buffer is the single truth for this min == max window.
    fn map_window(&mut self, qh: &QueueHandle<State>) {
        let (Some(comp), Some(shm), Some(wm)) = (&self.comp, &self.shm, &self.wm) else {
            return;
        };
        let surf = comp.create_surface(qh, ());
        let xsd = wm.get_xdg_surface(&surf, qh, ());
        let top = xsd.get_toplevel(qh, ());
        top.set_app_id(self.title.clone());
        top.set_title(self.title.clone());
        top.set_min_size(self.width, self.height);
        top.set_max_size(self.width, self.height); // min == max: the dialog shape
        self.surf = Some(surf);
        let width = u32::try_from(self.width).unwrap_or(1);
        let height = u32::try_from(self.height).unwrap_or(1);
        let count = (width * height) as usize;
        // the fd lives with the process; one fixture window, no reuse. A flat
        // 0xff202030 ARGB8888 pixmap, stride = w*4.
        let pixel = 0xff_20_20_30u32.to_ne_bytes();
        let mut pixels = Vec::with_capacity(count * 4);
        for _ in 0..count {
            pixels.extend_from_slice(&pixel);
        }
        let fd = memfd_create("fixwin", MemfdFlags::empty()).unwrap_or_else(|err| {
            eprintln!("fixwin: memfd: {err}");
            std::process::exit(1);
        });
        let mut offset = 0;
        while offset < pixels.len() {
            let n = match write(&fd, &pixels[offset..]) {
                Ok(n) => n,
                Err(err) => {
                    eprintln!("fixwin: pixel write: {err}");
                    std::process::exit(1);
                }
            };
            offset += n;
        }
        let size = i32::try_from(pixels.len()).unwrap_or(i32::MAX);
        let pool = shm.create_pool(fd.as_fd(), size, qh, ());
        let buf = pool.create_buffer(
            0,
            self.width,
            self.height,
            self.width * 4,
            wl_shm::Format::Argb8888,
            qh,
            (),
        );
        pool.destroy();
        self.buf = Some(buf);
        let (Some(surf), Some(b)) = (self.surf.as_ref(), self.buf.as_ref()) else {
            return;
        };
        surf.attach(Some(b), 0, 0);
        surf.commit();
    }
}

impl Dispatch<wl_registry::WlRegistry, ()> for State {
    fn event(
        state: &mut State,
        registry: &wl_registry::WlRegistry,
        event: wl_registry::Event,
        _data: &(),
        _conn: &Connection,
        qh: &QueueHandle<Self>,
    ) {
        if let wl_registry::Event::Global {
            name, interface, ..
        } = event
        {
            match interface.as_str() {
                "wl_compositor" if state.comp.is_none() => {
                    state.comp =
                        Some(registry.bind::<wl_compositor::WlCompositor, _, _>(name, 4, qh, ()));
                }
                "wl_shm" if state.shm.is_none() => {
                    state.shm = Some(registry.bind::<wl_shm::WlShm, _, _>(name, 1, qh, ()));
                }
                "xdg_wm_base" if state.wm.is_none() => {
                    state.wm = Some(registry.bind::<XdgWmBase, _, _>(name, 1, qh, ()));
                }
                _ => {}
            }
        }
    }
}

impl Dispatch<XdgWmBase, ()> for State {
    fn event(
        _state: &mut State,
        proxy: &XdgWmBase,
        event: XdgWmBaseEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
        if let XdgWmBaseEvent::Ping { serial } = event {
            proxy.pong(serial);
        }
    }
}

// The C fixture never acks configure (no xdg_surface configure listener, no
// toplevel configure action): the first committed buffer is the single truth
// for this min == max window, and acking changes what the compositor's
// placement sees. The port must stay byte-identical in protocol behavior.
impl Dispatch<XdgSurface, ()> for State {
    fn event(
        _state: &mut State,
        _proxy: &XdgSurface,
        _event: XdgSurfaceEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
    }
}

impl Dispatch<XdgToplevel, ()> for State {
    fn event(
        state: &mut State,
        _proxy: &XdgToplevel,
        event: XdgToplevelEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
        if let XdgToplevelEvent::Close = event {
            state.done = true;
        }
    }
}

// no handlers for these: the fixture never reacts to them
delegate_noop!(State: ignore wl_compositor::WlCompositor);
delegate_noop!(State: ignore wl_shm::WlShm);
delegate_noop!(State: ignore wl_shm_pool::WlShmPool);
delegate_noop!(State: ignore wl_buffer::WlBuffer);
delegate_noop!(State: ignore wl_surface::WlSurface);

fn main() {
    let mut args = std::env::args().skip(1);
    let width: i32 = args
        .next()
        .and_then(|arg| arg.parse().ok())
        .unwrap_or(310)
        .clamp(1, i32::MAX);
    let height: i32 = args
        .next()
        .and_then(|arg| arg.parse().ok())
        .unwrap_or(360)
        .clamp(1, i32::MAX);
    let title = args.next().unwrap_or_else(|| "fixwin".into());

    let Ok(conn) = Connection::connect_to_env() else {
        std::process::exit(1);
    };
    let mut queue = conn.new_event_queue();
    let qh = queue.handle();
    conn.display().get_registry(&qh, ());

    let mut state = State::new(width, height, title);
    // C drains the registry (dispatch + roundtrip) before creating anything
    while !state.ready() {
        if queue.blocking_dispatch(&mut state).is_err() {
            std::process::exit(1);
        }
    }
    if queue.roundtrip(&mut state).is_err() {
        std::process::exit(1);
    }
    state.map_window(&qh);
    while !state.done {
        if let Err(err) = queue.blocking_dispatch(&mut state) {
            eprintln!("fixwin: dispatch: {err}");
            break;
        }
    }
    let _ = conn.flush();
}
